// three/index.js
export const MathUtils = {
  degToRad: deg => deg * Math.PI / 180,
  radToDeg: rad => rad * 180 / Math.PI
};

export class Vector3 {
  constructor(x = 0, y = 0, z = 0) {
    this.x = x;
    this.y = y;
    this.z = z;
  }

  set(x, y, z) {
    this.x = x;
    this.y = y;
    this.z = z;
    return this;
  }

  copy(v) {
    this.x = v.x;
    this.y = v.y;
    this.z = v.z;
    return this;
  }

  clone() {
    return new Vector3(this.x, this.y, this.z);
  }

  add(v) {
    this.x += v.x;
    this.y += v.y;
    this.z += v.z;
    return this;
  }

  sub(v) {
    this.x -= v.x;
    this.y -= v.y;
    this.z -= v.z;
    return this;
  }

  multiplyScalar(s) {
    this.x *= s;
    this.y *= s;
    this.z *= s;
    return this;
  }

  lengthSq() {
    return this.x * this.x + this.y * this.y + this.z * this.z;
  }

  length() {
    return Math.sqrt(this.lengthSq());
  }

  normalize() {
    const len = this.length();
    if (len > 0) this.multiplyScalar(1 / len);
    return this;
  }

  distanceTo(v) {
    return Math.sqrt(
      (this.x - v.x) ** 2 +
      (this.y - v.y) ** 2 +
      (this.z - v.z) ** 2
    );
  }

  cross(v) {
    return this.crossVectors(this, v);
  }

  crossVectors(a, b) {
    const x = a.y * b.z - a.z * b.y;
    const y = a.z * b.x - a.x * b.z;
    const z = a.x * b.y - a.y * b.x;
    this.x = x;
    this.y = y;
    this.z = z;
    return this;
  }

  applyAxisAngle(axis, angle) {
    if (Math.abs(axis.y) === 1 && axis.x === 0 && axis.z === 0) {
      const c = Math.cos(angle);
      const s = Math.sin(angle) * Math.sign(axis.y);
      const x = this.x * c + this.z * s;
      const z = this.z * c - this.x * s;
      this.x = x;
      this.z = z;
    }
    return this;
  }

  project() {
    return this;
  }

  applyQuaternion() {
    return this;
  }
}

class Object3D {
  constructor() {
    this.children = [];
    this.position = new Vector3();
    this.rotation = { x: 0, y: 0, z: 0 };
    this.scale = new Vector3(1, 1, 1);
    this.scale.setScalar = s => {
      this.scale.set(s, s, s);
      return this.scale;
    };
    this.visible = true;
    this.userData = {};
  }

  add(...children) {
    children.forEach(child => {
      if (child) {
        child.parent = this;
        this.children.push(child);
      }
    });
  }

  remove(child) {
    this.children = this.children.filter(c => c !== child);
    if (child) child.parent = null;
  }

  traverse(callback) {
    callback(this);
    this.children.forEach(child => child.traverse ? child.traverse(callback) : callback(child));
  }

  setRotationFromEuler(euler) {
    this.rotation = euler;
  }

  lookAt() {}
}

export class Group extends Object3D {}
export class Scene extends Object3D {}

export class Mesh extends Object3D {
  constructor(geometry = null, material = null) {
    super();
    this.geometry = geometry;
    this.material = material;
    this.isMesh = true;
  }
}

export class Sprite extends Object3D {
  constructor(material = null) {
    super();
    this.material = material;
  }
}

export class Line extends Object3D {
  constructor(geometry = null, material = null) {
    super();
    this.geometry = geometry;
    this.material = material;
  }
}

export class ArrowHelper extends Object3D {
  constructor(dir = new Vector3(), origin = new Vector3(), length = 1, color = 0xffffff) {
    super();
    this.dir = dir;
    this.position.copy(origin);
    this.length = length;
    this.color = color;
  }

  setDirection(dir) {
    this.dir = dir.clone ? dir.clone() : dir;
  }
}

export class BufferGeometry {
  constructor() {
    this.points = [];
  }

  setFromPoints(points) {
    this.points = points;
    return this;
  }

  dispose() {}
}

export class LineBasicMaterial {
  constructor(options = {}) {
    Object.assign(this, options);
  }

  dispose() {}
}

export class MeshBasicMaterial {
  constructor(options = {}) {
    Object.assign(this, options);
  }

  dispose() {}
}

export class MeshPhongMaterial extends MeshBasicMaterial {}
export class MeshStandardMaterial extends MeshBasicMaterial {}
export class SpriteMaterial extends MeshBasicMaterial {}
export class ShaderMaterial extends MeshBasicMaterial {}

export class SphereGeometry {
  constructor(radius = 1) {
    this.parameters = { radius };
  }

  dispose() {}
}

export class BoxGeometry extends SphereGeometry {}

export class Box3 {
  setFromObject() {
    return this;
  }

  getCenter(target) {
    target.set(0, 0, 0);
    return target;
  }

  getSize(target) {
    target.set(1, 1, 1);
    return target;
  }
}

export class Path {
  absarc() {
    return this;
  }

  getPoints(count = 0) {
    return Array.from({ length: count + 1 }, (_, i) => new Vector3(i, 0, 0));
  }
}

export class Color {
  constructor(value = 0xffffff) {
    this.value = value;
  }

  set(value) {
    this.value = value;
    return this;
  }
}

export class TextureLoader {
  load(url, onLoad) {
    const texture = { url, isTexture: true, dispose() {} };
    if (onLoad) setTimeout(() => onLoad(texture), 0);
    return texture;
  }
}

export class CanvasTexture {
  constructor(canvas) {
    this.canvas = canvas;
    this.isTexture = true;
  }
}

export class DirectionalLight extends Object3D {
  constructor(color = 0xffffff, intensity = 1) {
    super();
    this.color = color;
    this.intensity = intensity;
  }
}

export class AmbientLight extends DirectionalLight {}

export class Euler {
  constructor(x = 0, y = 0, z = 0, order = 'XYZ') {
    this.x = x;
    this.y = y;
    this.z = z;
    this.order = order;
  }
}

export class Quaternion {
  setFromEuler() {
    return this;
  }
}

export const AdditiveBlending = 'AdditiveBlending';
export const LinearFilter = 'LinearFilter';
export const SRGBColorSpace = 'srgb';

export default {
  MathUtils,
  Vector3,
  Group,
  Scene,
  Mesh,
  Sprite,
  Line,
  ArrowHelper,
  BufferGeometry,
  LineBasicMaterial,
  MeshBasicMaterial,
  MeshPhongMaterial,
  MeshStandardMaterial,
  SpriteMaterial,
  ShaderMaterial,
  SphereGeometry,
  BoxGeometry,
  Box3,
  Path,
  Color,
  TextureLoader,
  CanvasTexture,
  DirectionalLight,
  AmbientLight,
  Euler,
  Quaternion,
  AdditiveBlending,
  LinearFilter,
  SRGBColorSpace
};
